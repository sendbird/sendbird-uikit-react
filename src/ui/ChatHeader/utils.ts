import { GroupChannel } from '@sendbird/chat/groupChannel';

export const getChannelTitle = (channel: GroupChannel, currentUserId: string, stringSet: { [key: string]: string }): string => {
  if (!channel?.name && !channel?.members) {
    return stringSet.NO_TITLE;
  }
  if (channel?.name && channel.name !== 'Group Channel') {
    return channel.name;
  }

  if (channel?.members?.length === 1) {
    return stringSet.NO_MEMBERS;
  }

  return (channel?.members || [])
    .filter(({ userId }) => userId !== currentUserId)
    .map(({ nickname }) => (nickname || stringSet.NO_NAME))
    .join(', ');
};
