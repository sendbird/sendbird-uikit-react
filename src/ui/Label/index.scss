@import '../../styles/variables';

.sendbird-label {
  -webkit-user-select: none;
  -webkit-touch-callout: none;
}

[class*=sendbird-label] {
  font-family: var(--sendbird-font-family-default);
}

.sendbird-label--h-1 {
  font-size: 20px;
  font-weight: 600;
  font-stretch: normal;
  font-style: normal;
  line-height: 1.4;
  letter-spacing: -0.2px;
}

.sendbird-label--h-2 {
  font-size: 18px;
  font-weight: 600;
  font-stretch: normal;
  font-style: normal;
  line-height: normal;
  letter-spacing: -0.2px;
}

.sendbird-label--subtitle-1 {
  font-size: 16px;
  font-weight: normal;
  font-stretch: normal;
  font-style: normal;
  line-height: 1.38;
  letter-spacing: -0.2px;
}

.sendbird-label--subtitle-2 {
  font-size: 14px;
  font-weight: 500;
  font-stretch: normal;
  font-style: normal;
  line-height: 1.14;
  letter-spacing: -0.2px;
}

.sendbird-label--body-1 {
 font-size: 14px;
 font-weight: normal;
 font-stretch: normal;
 font-style: normal;
 line-height: 1.43;
 letter-spacing: normal;
}

.sendbird-label--body-2 {
  font-size: 12px;
  font-weight: normal;
  font-stretch: normal;
  font-style: normal;
  line-height: 1.33;
  letter-spacing: normal;
}

.sendbird-label--button-1 {
  font-size: 14px;
  font-weight: 600;
  font-stretch: normal;
  font-style: normal;
  line-height: 1.43;
  letter-spacing: normal;
}

.sendbird-label--button-2 {
  font-size: 14px;
  font-weight: normal;
  font-stretch: normal;
  font-style: normal;
  line-height: 1.43;
  letter-spacing: normal;
}

.sendbird-label--caption-1 {
 font-size: 14px;
 font-weight: 600;
 font-stretch: normal;
 font-style: normal;
 line-height: 1.43;
 letter-spacing: normal;
}

.sendbird-label--caption-2 {
 font-size: 12px;
 font-weight: bold;
 font-style: normal;
 line-height: 1;
 letter-spacing: normal;
}

.sendbird-label--caption-3 {
  font-size: 12px;
  font-weight: normal;
  font-stretch: normal;
  font-style: normal;
  line-height: 1;
  letter-spacing: normal;
}

.sendbird-label--color-onbackground-1 {
  @include themed() {
    color: t(on-bg-1);
  }
}

.sendbird-label--color-onbackground-2 {
  @include themed() {
    color: t(on-bg-2);
  }
}

.sendbird-label--color-onbackground-3 {
  @include themed() {
    color: t(on-bg-3);
  }
}

.sendbird-label--color-onbackground-4 {
  @include themed() {
    color: t(on-bg-4);
  }
}

.sendbird-label--color-oncontent-1 {
  @include themed() {
    color: t(on-content-1);
  }
}

.sendbird-label--color-oncontent-2 {
  @include themed() {
    color: t(on-content-2);
  }
}

.sendbird-label--color-primary {
  @include themed() {
    color: t(primary-3);
  }
}

.sendbird-label--color-error {
  @include themed() {
    color: t(error--3-2);
  }
}

.sendbird-label--color-secondary-3 {
  @include themed() {
    color: t(secondary-3);
  }
}
