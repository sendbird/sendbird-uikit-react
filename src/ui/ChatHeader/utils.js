import formatDistanceToNowStrict from 'date-fns/formatDistanceToNowStrict';
import { LabelStringSet } from '../Label';

export const prettyDate = (date) => formatDistanceToNowStrict(date, { addSuffix: true });

export const getChannelTitle = (channel = {}, currentUserId, stringSet = LabelStringSet) => {
  if (!channel?.name && !channel?.members) {
    return stringSet.NO_TITLE;
  }
  if (channel?.name && channel.name !== 'Group Channel') {
    return channel.name;
  }

  if (channel?.members?.length === 1) {
    return stringSet.NO_MEMBERS;
  }

  return (channel?.members || [])
    .filter(({ userId }) => userId !== currentUserId)
    .map(({ nickname }) => (nickname || stringSet.NO_NAME))
    .join(', ');
};
