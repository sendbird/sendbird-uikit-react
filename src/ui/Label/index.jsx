import React from 'react';
import PropTypes from 'prop-types';

import './index.scss';
import { Typography, Colors } from './types';
import { changeTypographyToClassName, changeColorToClassName } from './utils';
import getStringSet from './stringSet';

export default function Label({
  className,
  type,
  color,
  children,
}) {
  return (
    <span
      className={[
        ...(Array.isArray(className) ? className : [className]),
        'sendbird-label',
        type ? changeTypographyToClassName(type) : '',
        color ? changeColorToClassName(color) : '',
      ].join(' ')}
    >
      {children}
    </span>
  );
}

Label.propTypes = {
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]),
  type: PropTypes.oneOf([
    'H_1',
    'H_2',
    'SUBTITLE_1',
    'SUBTITLE_2',
    'BODY_1',
    'BODY_2',
    'BUTTON_1',
    'BUTTON_2',
    'CAPTION_1',
    'CAPTION_2',
    'CAPTION_3',
  ]),
  color: PropTypes.oneOf([
    'ONBACKGROUND_1',
    'ONBACKGROUND_2',
    'ONBACKGROUND_3',
    'ONBACKGROUND_4',
    'ONCONTENT_1',
    'ONCONTENT_2',
    'PRIMARY',
    'ERROR',
    'SECONDARY_3',
  ]),
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.element,
    PropTypes.any,
  ]),
};

Label.defaultProps = {
  className: [],
  type: '',
  color: '',
  children: null,
};

const LabelTypography = Typography;
const LabelColors = Colors;
const LabelStringSet = getStringSet('en');
export { LabelTypography, LabelColors, LabelStringSet };
