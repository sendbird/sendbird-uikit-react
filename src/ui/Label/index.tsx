import React, {RefObject} from 'react';

import './index.scss';
import { Typography, Colors } from './types';
import { changeTypographyToClassName, changeColorToClassName } from './utils';
import getStringSet from './stringSet';
import { ObjectValues } from '../../utils/typeHelpers/objectValues';
import {ThreadReplies} from '../ThreadReplies';

type LabelProps = {
  className?: string | string[];
  type?: ObjectValues<typeof Typography>;
  color?: ObjectValues<typeof Colors>;
  children?: React.ReactNode;
};
export function Label(
  {
    className = [],
    type,
    color,
    children = null,
  }: LabelProps,
  ref?: RefObject<HTMLDivElement>,
) {
  return (
    // Donot make this into div
    // Mention uses Label. If we use div, it would break the mention detection on Paste
    // https://github.com/sendbird/sendbird-uikit-react/pull/479
    <span
      className={[
        ...(Array.isArray(className) ? className : [className]),
        'sendbird-label',
        changeTypographyToClassName(type),
        changeColorToClassName(color),
      ].join(' ')}
      ref={ref}
    >
      {children}
    </span>
  );
}

const LabelTypography = Typography;
const LabelColors = Colors;
const LabelStringSet = getStringSet('en');
export { LabelTypography, LabelColors, LabelStringSet };

export default React.forwardRef(Label);