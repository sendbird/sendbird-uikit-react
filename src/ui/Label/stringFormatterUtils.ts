import {CoreMessageType, isMultipleFilesMessage} from "../../utils";
import {match} from "ts-pattern";
import {MultipleFilesMessage} from "@sendbird/chat/message";

export function getModalDeleteMessageTitle(stringSet: Record<string, string>, message: CoreMessageType): string {
  return match(message)
    .when(isMultipleFilesMessage, () => {
      const filesCount: number = (message as MultipleFilesMessage).fileInfoList.length;
      return `Do you want to delete all ${filesCount} photos?`;
    })
    .otherwise(() => {
      return stringSet.MODAL__DELETE_MESSAGE__TITLE;
    });
}
