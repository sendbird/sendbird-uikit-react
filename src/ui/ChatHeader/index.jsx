// delete
import React, { useContext } from 'react';
import PropTypes from 'prop-types';

import './index.scss';
import * as utils from './utils';

import { LocalizationContext } from '../../lib/LocalizationContext';
import Label, { LabelTypography, LabelColors } from '../Label';
import Icon, { IconTypes, IconColors } from '../Icon';
import IconButton from '../IconButton';
import ChannelAvatar from '../ChannelAvatar/index';

const noop = () => { };
export default function ChatHeader(props) {
  const {
    currentGroupChannel,
    currentUser,
    title,
    subTitle,
    isMuted,
    theme,
    showSearchIcon,
    onSearchClick,
    onActionClick,
  } = props;
  const { userId } = currentUser;
  const { stringSet } = useContext(LocalizationContext);
  return (
    <div className="sendbird-chat-header">
      <div className="sendbird-chat-header__left">
        <ChannelAvatar
          theme={theme}
          channel={currentGroupChannel}
          userId={userId}
          height={32}
          width={32}
        />
        <Label
          className="sendbird-chat-header__left__title"
          type={LabelTypography.H_2}
          color={LabelColors.ONBACKGROUND_1}
        >
          {title || utils.getChannelTitle(currentGroupChannel, userId, stringSet)}
        </Label>
        <Label
          className="sendbird-chat-header__left__subtitle"
          type={LabelTypography.BODY_1}
          color={LabelColors.ONBACKGROUND_2}
        >
          {subTitle}
        </Label>
      </div>
      <div className="sendbird-chat-header__right">
        {
          ((typeof isMuted === 'string' && isMuted === 'true') || (typeof isMuted === 'boolean' && isMuted))
            && (
              <Icon
                className="sendbird-chat-header__right__mute"
                type={IconTypes.NOTIFICATIONS_OFF_FILLED}
                width="24px"
                height="24px"
              />
            )
        }
        {
          showSearchIcon && (
            <IconButton
              className="sendbird-chat-header__right__search"
              width="32px"
              height="32px"
              onClick={onSearchClick}
            >
              <Icon
                type={IconTypes.SEARCH}
                fillColor={IconColors.PRIMARY}
                width="24px"
                height="24px"
              />
            </IconButton>
          )
        }
        <IconButton
          className="sendbird-chat-header__right__info"
          width="32px"
          height="32px"
          onClick={onActionClick}
        >
          <Icon
            type={IconTypes.INFO}
            fillColor={IconColors.PRIMARY}
            width="24px"
            height="24px"
          />
        </IconButton>
      </div>
    </div>
  );
}

ChatHeader.propTypes = {
  /** Type: GroupChannel */
  currentGroupChannel: PropTypes.shape({
    members: PropTypes.arrayOf(PropTypes.shape({})),
    coverUrl: PropTypes.string,
  }),
  /** Type: User */
  currentUser: PropTypes.shape({
    userId: PropTypes.string,
  }),
  title: PropTypes.string,
  subTitle: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.string,
  ]),
  isMuted: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.bool,
  ]),
  theme: PropTypes.string,
  showSearchIcon: PropTypes.bool,
  onSearchClick: PropTypes.func,
  /** For clicking the info button */
  onActionClick: PropTypes.func,
};

ChatHeader.defaultProps = {
  currentGroupChannel: {},
  currentUser: {},
  title: '',
  subTitle: '',
  isMuted: false,
  theme: 'light',
  showSearchIcon: false,
  onSearchClick: noop,
  onActionClick: noop,
};
