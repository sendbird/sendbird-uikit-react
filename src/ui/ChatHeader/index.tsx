// will be deprecated
import React, { ReactElement, useContext, useEffect } from 'react';
import { User } from '@sendbird/chat';
import { GroupChannel } from '@sendbird/chat/groupChannel';

import './index.scss';
import * as utils from './utils';

import { LocalizationContext } from '../../lib/LocalizationContext';
import Label, { LabelTypography, LabelColors } from '../Label';
import Icon, { IconTypes, IconColors } from '../Icon';
import IconButton from '../IconButton';
import ChannelAvatar from '../ChannelAvatar/index';

export interface ChatHeaderProps {
  currentGroupChannel?: GroupChannel;
  currentUser?: User;
  title?: string;
  subTitle?: string;
  isMuted?: boolean;
  theme?: string;
  showSearchIcon?: boolean;
  onSearchClick?: () => void;
  onActionClick?: () => void;
}

export default function ChatHeader({
  currentGroupChannel,
  currentUser,
  title = '',
  subTitle = '',
  isMuted = false,
  theme = 'light',
  showSearchIcon = false,
  onSearchClick = () => {/* noop */},
  onActionClick = () => {/* noop */},
}: ChatHeaderProps): ReactElement {
  const userId = currentUser?.userId;
  const { stringSet } = useContext(LocalizationContext);
  useEffect(() => {
    // eslint-disable-next-line no-console
    console.warn('SendbirdUIKit: ChatHeader will be deprecated in the next minor version. Recommend to use "@sendbird/uikit-react/Channtl/components/ChannelHeader" instead.');
  }, []);
  return (
    <div className="sendbird-chat-header">
      <div className="sendbird-chat-header__left">
        <ChannelAvatar
          theme={theme}
          channel={currentGroupChannel}
          userId={userId}
          height={32}
          width={32}
        />
        <Label
          className="sendbird-chat-header__left__title"
          type={LabelTypography.H_2}
          color={LabelColors.ONBACKGROUND_1}
        >
          {title || utils.getChannelTitle(currentGroupChannel, userId, stringSet)}
        </Label>
        <Label
          className="sendbird-chat-header__left__subtitle"
          type={LabelTypography.BODY_1}
          color={LabelColors.ONBACKGROUND_2}
        >
          {subTitle}
        </Label>
      </div>
      <div className="sendbird-chat-header__right">
        {
          ((typeof isMuted === 'string' && isMuted === 'true') || (typeof isMuted === 'boolean' && isMuted))
            && (
              <Icon
                className="sendbird-chat-header__right__mute"
                type={IconTypes.NOTIFICATIONS_OFF_FILLED}
                width="24px"
                height="24px"
              />
            )
        }
        {
          showSearchIcon && (
            <IconButton
              className="sendbird-chat-header__right__search"
              width="32px"
              height="32px"
              onClick={onSearchClick}
            >
              <Icon
                type={IconTypes.SEARCH}
                fillColor={IconColors.PRIMARY}
                width="24px"
                height="24px"
              />
            </IconButton>
          )
        }
        <IconButton
          className="sendbird-chat-header__right__info"
          width="32px"
          height="32px"
          onClick={onActionClick}
        >
          <Icon
            type={IconTypes.INFO}
            fillColor={IconColors.PRIMARY}
            width="24px"
            height="24px"
          />
        </IconButton>
      </div>
    </div>
  );
}
