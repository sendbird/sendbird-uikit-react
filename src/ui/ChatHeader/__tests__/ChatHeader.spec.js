import React from 'react';
import { render, screen } from '@testing-library/react';

import ChatHeader from "../index";
import { LabelStringSet } from '../../Label';
const title = "Headline 2";
const subTitle = "Body text 2";
jest.useFakeTimers();

describe('ui/ChatHeader', () => {
  it('should do a snapshot test of the ChatHeader DOM', function () {
    const { asFragment } = render(
      <ChatHeader
        title={title}
        subTitle={subTitle}
        isActive="true"
        isMuted="true"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should have each components in the ChatHeader', function () {
    const { container } = render(
      <ChatHeader
        title={title}
        subTitle={subTitle}
        isActive="true"
        isMuted="true"
      />
    );

    expect(screen.getByText(title).className).toContain('sendbird-chat-header__left__title');
    expect(screen.getByText(subTitle).className).toContain('sendbird-chat-header__left__subtitle');
    expect(container.querySelector('.sendbird-chat-header__right__mute').className).toContain('sendbird-icon');
    expect(container.querySelector('.sendbird-chat-header__right__info').className).toContain('sendbird-iconbutton');
  });

  it('should not have components in the ChatHeader', function () {
    const { container } = render(<ChatHeader />);

    expect(container.querySelector('.sendbird-chat-header__left__title').textContent).toBe(LabelStringSet.NO_TITLE);
    expect(container.querySelector('.sendbird-chat-header__left__subtitle').textContent).toBe('');
    expect(container.querySelector('.sendbird-chat-header__right__mute')).toBeNull();
    expect(container.querySelector('.sendbird-chat-header__right__info').className).toContain('sendbird-iconbutton');
  });
});
