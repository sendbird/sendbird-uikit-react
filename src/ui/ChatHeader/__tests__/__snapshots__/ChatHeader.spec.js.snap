// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatHeader should do a snapshot test of the ChatHeader DOM 1`] = `
<div
  className="sendbird-chat-header"
>
  <div
    className="sendbird-chat-header__left"
  >
    <div
      className="sendbird-chat-header--avatar--group-channel sendbird-avatar"
      role="button"
      style={
        Object {
          "height": "32px",
          "width": "32px",
        }
      }
      tabIndex={0}
    >
      <div
        className="sendbird-avatar-img sendbird-image-renderer"
        style={
          Object {
            "height": "32px",
            "maxWidth": "400px",
            "minWidth": "32px",
            "width": "100%",
          }
        }
      >
        <div
          className="sendbird-image-renderer__image"
          style={
            Object {
              "backgroundImage": "url()",
              "backgroundPosition": "center",
              "backgroundRepeat": "no-repeat",
              "backgroundSize": "cover",
              "borderRadius": null,
              "height": "32px",
              "maxWidth": "400px",
              "minWidth": "32px",
              "position": "absolute",
              "width": "100%",
            }
          }
        />
        <img
          alt=""
          className="sendbird-image-renderer__hidden-image-loader"
          onError={[Function]}
          onLoad={[Function]}
          src=""
        />
      </div>
    </div>
    <span
      className="sendbird-chat-header__left__title sendbird-label sendbird-label--h-2 sendbird-label--color-onbackground-1"
    >
      Headline 2
    </span>
    <span
      className="sendbird-chat-header__left__subtitle sendbird-label sendbird-label--body-1 sendbird-label--color-onbackground-2"
    >
      Body text 2
    </span>
  </div>
  <div
    className="sendbird-chat-header__right"
  >
    <div
      className="sendbird-chat-header__right__mute sendbird-icon sendbird-icon-notifications-off-filled "
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      style={
        Object {
          "height": "24px",
          "minHeight": "24px",
          "minWidth": "24px",
          "width": "24px",
        }
      }
      tabIndex="0"
    >
      <test-file-stub />
    </div>
    <button
      className="sendbird-chat-header__right__info sendbird-iconbutton "
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      style={
        Object {
          "height": "32px",
          "width": "32px",
        }
      }
      type="button"
    >
      <span
        className="sendbird-iconbutton__inner"
      >
        <div
          className=" sendbird-icon sendbird-icon-info sendbird-icon-color--primary"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          style={
            Object {
              "height": "24px",
              "minHeight": "24px",
              "minWidth": "24px",
              "width": "24px",
            }
          }
          tabIndex="0"
        >
          <test-file-stub />
        </div>
      </span>
    </button>
  </div>
</div>
`;
